networkconfig.py

```diff
@@ -3,201 +3,304 @@
 import ure
 import time
 
+ap_ssid = "WifiManager"
+ap_password = "tayfunulu"
+ap_authmode = 3  # WPA2
+
+NETWORK_PROFILES = 'wifi.dat'
+
 wlan_ap = network.WLAN(network.AP_IF)
 wlan_sta = network.WLAN(network.STA_IF)
 
-# setup for ssid 
-ssid_name= "WifiManager"
-ssid_password = "tayfunulu"
-
 server_socket = None
 
-def do_connect(ntwrk_ssid, netwrk_pass):
-	sta_if = network.WLAN(network.STA_IF)
-	sta_if.active(True)
-	if not sta_if.isconnected():
-		print('try to connect : '+ntwrk_ssid+' network...')
-		sta_if.active(True)
-		sta_if.connect(ntwrk_ssid, netwrk_pass)
-		a=0
-		while not sta_if.isconnected() | (a > 99) :
-			time.sleep(0.1)
-			a+=1
-			print('.', end='')
-		if sta_if.isconnected():
-			print('\nConnected. Network config:', sta_if.ifconfig())
-			return (True)
-		else : 
-			print('\nProblem. Not Connected to :'+ntwrk_ssid)
-			return (False)
+
+def get_connection():
+    """return a working WLAN(STA_IF) instance or None"""
+
+    # First check if there already is any connection:
+    if wlan_sta.isconnected():
+        return wlan_sta
+
+    connected = False
+    try:
+        # ESP connecting to WiFi takes time, wait a bit and try again:
+        time.sleep(3)
+        if wlan_sta.isconnected():
+            return wlan_sta
+
+        # Read known network profiles from file
+        profiles = read_profiles()
+
+        # Search WiFis in range
+        wlan_sta.active(True)
+        networks = wlan_sta.scan()
+
+        AUTHMODE = {0: "open", 1: "WEP", 2: "WPA-PSK", 3: "WPA2-PSK", 4: "WPA/WPA2-PSK"}
+        for ssid, bssid, channel, rssi, authmode, hidden in sorted(networks, key=lambda x: x[3], reverse=True):
+            ssid = ssid.decode('utf-8')
+            encrypted = authmode > 0
+            print("ssid: %s chan: %d rssi: %d authmode: %s" % (ssid, channel, rssi, AUTHMODE.get(authmode, '?')))
+            if encrypted:
+                if ssid in profiles:
+                    password = profiles[ssid]
+                    connected = do_connect(ssid, password)
+                else:
+                    print("skipping unknown encrypted network")
+            else:  # open
+                connected = do_connect(ssid, None)
+            if connected:
+                break
+
+    except OSError as e:
+        print("exception", str(e))
```

Comment:
the "pass" that was here first swalled exceptions, of course.
including the one that the STA was not active...